### DISCUSSION TITLE
TorchAudio feasibility

### DISCUSSION THREAD

**Discussion by GeorgeS2019** [xamarin/TorchSharp#598] [Ideas]
# PyTorch audio: TorchAudio

- [ ] [Audio I/O](https://pytorch.org/audio/stable/tutorials/audio_io_tutorial.html)
- [x] [Audio Resampling](https://pytorch.org/tutorials/beginner/audio_resampling_tutorial.html)
- [ ] [Audio Data Augmentation](https://pytorch.org/tutorials/beginner/audio_data_augmentation_tutorial.html)
- [x] [Audio Feature Extractions](https://pytorch.org/tutorials/beginner/audio_feature_extractions_tutorial.html)
- [ ] [Audio Feature Augmentation](https://pytorch.org/tutorials/beginner/audio_feature_augmentation_tutorial.html)
- [x] [Audio Datasets](https://pytorch.org/tutorials/beginner/audio_datasets_tutorial.html)
- [x]  [Speech Recognition with Wav2Vec2](https://pytorch.org/tutorials/intermediate/speech_recognition_pipeline_tutorial.html)
- [x] [Speech Command Classification with torchaudio](https://pytorch.org/tutorials/intermediate/speech_command_classification_with_torchaudio_tutorial.html)
- [ ]  [Text-to-speech with torchaudio](https://pytorch.org/tutorials/intermediate/text_to_speech_with_torchaudio.html)
- [x] [Text-to-speech with torchaudio: tacotron2](https://pytorch.org/audio/stable/tutorials/tacotron2_pipeline_tutorial.html)

@[kaiidams](https://github.com/kaiidams) could you please check and perhaps **organize** what you have done could support .NET community to work towards TorchAudio using TorchSharp.

**Comment by GeorgeS2019** [comment]
@luisquintanilla I hope you could find time to look at some of the deep voice AI contributions for .NET community by @kaiidams.  He is in my view a suitable candidate for one of your monthly ML.NET standup community programs.

- [Voice100Godot](https://github.com/kaiidams/Voice100Godot)
- [TransferLearningAudio](https://github.com/kaiidams/TransferLearningAudio)
- [Voice100AndroidApp](https://github.com/kaiidams/Voice100AndroidApp)
- [NeMoOnnxSharp](https://github.com/kaiidams/NeMoOnnxSharp)

**Reply by luisquintanilla** [reply]
Hi @GeorgeS2019 

This is great. Thanks for sharing! I'll take a look and we'd be happy to have @kaiidams on 🙂

**Reply by kaiidams** [reply]
@GeorgeS2019 Thank you for introduction.
@luisquintanilla I am working on torch.stft/istft() which is missing and will be required for TorchAudio.
https://github.com/kaiidams/TorchSharp/commit/ada65452324ae7ee16d975bd4a2eb963564be4b6
I'll need next weekend to polish parameter conversions.

**Reply by NiklasGustafsson** [reply]
@kaiidams PR has been merged a couple of days ago, by the way.

**Comment by GeorgeS2019** [comment]
[first of many TourchAudio commits](https://github.com/dotnet/TorchSharp/tree/main/src/TorchSharp/TorchAudio)

**Comment by GeorgeS2019** [comment]
- [x] [TorchSharp: Audio Resampling](https://github.com/dotnet/TorchSharp/blob/main/src/TorchSharp/TorchAudio/ResamplingMethod.cs) on  [torchautio tutorial](https://pytorch.org/tutorials/beginner/audio_resampling_tutorial.html)
#### Resampling Overview
> To resample an audio waveform from one freqeuncy to another, you can use TorchSharp's torchaudio.transforms.[Resample()](https://github.com/dotnet/TorchSharp/blob/bc926e9fe5215303fa039e205dd1312c859f9a18/src/TorchSharp/TorchAudio/Transforms.cs#L185) or torchaudio.functional.[resample()](https://github.com/dotnet/TorchSharp/blob/bc926e9fe5215303fa039e205dd1312c859f9a18/src/TorchSharp/TorchAudio/Functional.cs#L165). transforms.Resample precomputes and caches the kernel used for resampling, while functional.resample computes it on the fly, so using torchaudio.transforms.Resample will result in a speedup when resampling multiple waveforms using the same parameters

- [x] [TorchSharp; Audio Datasets](https://github.com/dotnet/TorchSharp/commits/main/src/TorchSharp/TorchAudio/Datasets/YesnoDataset.cs) on  [torchautio tutorial](https://pytorch.org/tutorials/beginner/audio_datasets_tutorial.html)
 - [x] [TorchSharp: Speech Command Classification with torchaudio](https://github.com/dotnet/TorchSharp/blob/main/src/Examples/SpeechCommands.cs) based on this [PR](https://github.com/dotnet/TorchSharp/pull/649) on  [torchautio tutorial](https://pytorch.org/tutorials/intermediate/speech_command_classification_with_torchaudio_tutorial.html?highlight=torchaudio)
- #### Importing the Dataset
```python
from torchaudio.datasets import SPEECHCOMMANDS
```
[TorchSharp](https://github.com/dotnet/TorchSharp/blob/bc926e9fe5215303fa039e205dd1312c859f9a18/src/TorchSharp/TorchAudio/Datasets/SpeechCommandsDataset.cs#L105)
```cs
 public static torch.utils.data.Dataset<SpeechCommandsDatasetItem> SPEECHCOMMANDS
```
- #### Formatting the Data
```python
transform = torchaudio.transforms.Resample(orig_freq=sample_rate, new_freq=new_sample_rate)
```
[TorchSharp](https://github.com/dotnet/TorchSharp/blob/bc926e9fe5215303fa039e205dd1312c859f9a18/src/Examples/SpeechCommands.cs#L69)
```cs
var transform = torchaudio.transforms.Resample(_sample_rate, _new_sample_rate, device: device);
```

- #### [Define the Network](https://github.com/dotnet/TorchSharp/blob/bc926e9fe5215303fa039e205dd1312c859f9a18/src/Examples/SpeechCommands.cs#L220)
- #### [Training](https://github.com/dotnet/TorchSharp/blob/bc926e9fe5215303fa039e205dd1312c859f9a18/src/Examples/SpeechCommands.cs#L122) and [Testing](https://github.com/dotnet/TorchSharp/blob/bc926e9fe5215303fa039e205dd1312c859f9a18/src/Examples/SpeechCommands.cs#L160) the Network

**Comment by GeorgeS2019** [comment]
- [x] [TorchSharp: Audio Feature Extractions] based on [torchaudio tutorial](https://pytorch.org/tutorials/beginner/audio_feature_extractions_tutorial.html)
> torchaudio implements feature extractions commonly used in the audio domain. They are available in **TorchSharp**: torchaudio.[functional ](https://github.com/dotnet/TorchSharp/blob/main/src/TorchSharp/TorchAudio/Functional.cs)and torchaudio.[transforms](https://github.com/dotnet/TorchSharp/blob/main/src/TorchSharp/TorchAudio/Transforms.cs).

Possible [.NET Spectrogram](https://github.com/swharden/Spectrogram) plotting library

For .NET interactive notebook, we need this display supported in 
- [Plotly.NET](https://github.com/plotly/Plotly.NET/issues/282)

Which need support in 
- [plotly.js](https://github.com/plotly/plotly.js)
  > [Spectrogram feature request? ](https://github.com/plotly/plotly.js/issues/6273)

**Comment by GeorgeS2019** [comment]
- [x] [TEXT-TO-SPEECH WITH TACOTRON2](https://pytorch.org/audio/stable/tutorials/tacotron2_pipeline_tutorial.html) using [TorchSharp](https://github.com/dotnet/TorchSharp/pull/669)

**Comment by GeorgeS2019** [comment]
torchaudio.[function](https://github.com/dotnet/TorchSharp/blob/main/src/TorchSharp/TorchAudio/Functional.cs) implements features as standalone functions. They are stateless.

new [TorchSharp  torchaudio.function PR](https://github.com/dotnet/TorchSharp/pull/685)



![image](https://user-images.githubusercontent.com/49812372/183240395-549d74db-fea7-4af3-a883-142ea862d054.png)

**Comment by GeorgeS2019** [comment]
@kaiidams 
#### One of many TorchSharp's TorchAudio Goals: [Language is essential to human life](https://github.com/sillsdev/machine/issues/10#issuecomment-1225368911)

**Comment by GeorgeS2019** [comment]
Another TorchSharp Wav2Vec2 example
@whuanle

https://github.com/whuanle/Maomi.Torch/blob/main/examples/wav2vec2/Program.cs

**Comment by GeorgeS2019** [comment]
https://github.com/dotnet/TorchSharp/issues/1377#issuecomment-2730024841
